src/wp-includes/speculative-loading.php

```diff
@@ -0,0 +1,309 @@
+		if ( null === $configuration ) {
+			return array();
+		}
+	}
+
+	$mode      = $configuration['mode'];
+	$eagerness = $configuration['eagerness'];
+
+	$prefixer = new WP_URL_Pattern_Prefixer();
+
+	$base_href_exclude_paths = array(
+		$prefixer->prefix_path_pattern( '/wp-*.php', 'site' ),
+		$prefixer->prefix_path_pattern( '/wp-admin/*', 'site' ),
+		$prefixer->prefix_path_pattern( '/*', 'uploads' ),
+		$prefixer->prefix_path_pattern( '/*', 'content' ),
+		$prefixer->prefix_path_pattern( '/*', 'plugins' ),
+		$prefixer->prefix_path_pattern( '/*', 'template' ),
+		$prefixer->prefix_path_pattern( '/*', 'stylesheet' ),
+	);
+
+	/*
+	 * If pretty permalinks are enabled, exclude any URLs with query parameters.
+	 * Otherwise, exclude specifically the URLs with a `_wpnonce` query parameter.
+	 */
+	if ( get_option( 'permalink_structure' ) ) {
+		$base_href_exclude_paths[] = $prefixer->prefix_path_pattern( '/*\\?(.+)', 'home' );
+	} else {
+		$base_href_exclude_paths[] = $prefixer->prefix_path_pattern( '/*\\?*(^|&)_wpnonce=*', 'home' );
+	}
+
+	/**
+	 * Filters the paths for which speculative loading should be disabled.
+	 *
+	 * All paths should start in a forward slash, relative to the root document. The `*` can be used as a wildcard.
+	 * If the WordPress site is in a subdirectory, the exclude paths will automatically be prefixed as necessary.
+	 *
+	 * Note that WordPress always excludes certain path patterns such as `/wp-login.php` and `/wp-admin/*`, and those
+	 * cannot be modified using the filter.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @param string[] $href_exclude_paths Additional path patterns to disable speculative loading for.
+	 * @param string   $mode               Mode used to apply speculative loading. Either 'prefetch' or 'prerender'.
+	 */
+	$href_exclude_paths = (array) apply_filters( 'wp_speculation_rules_href_exclude_paths', array(), $mode );
+
+	// Ensure that:
+	// 1. There are no duplicates.
+	// 2. The base paths cannot be removed.
+	// 3. The array has sequential keys (i.e. array_is_list()).
+	$href_exclude_paths = array_values(
+		array_unique(
+			array_merge(
+				$base_href_exclude_paths,
+				array_map(
+					static function ( string $href_exclude_path ) use ( $prefixer ): string {
+						return $prefixer->prefix_path_pattern( $href_exclude_path );
+					},
+					$href_exclude_paths
+				)
+			)
+		)
+	);
+
+	$speculation_rules = new WP_Speculation_Rules();
+
+	$main_rule_conditions = array(
+		// Include any URLs within the same site.
+		array(
+			'href_matches' => $prefixer->prefix_path_pattern( '/*' ),
+		),
+		// Except for excluded paths.
+		array(
+			'not' => array(
+				'href_matches' => $href_exclude_paths,
+			),
+		),
+		// Also exclude rel=nofollow links, as certain plugins use that on their links that perform an action.
+		array(
+			'not' => array(
+				'selector_matches' => 'a[rel~="nofollow"]',
+			),
+		),
+		// Last but not least, exclude links that are explicitly marked to opt out.
```

Comment:
I'd remove this just to make it easier for our future selves to add more without forgetting to update this.
```suggestion
		// Exclude links that are explicitly marked to opt out.
```